### ISSUE TITLE
Movies recommendations Systems

### ISSUE THREAD

**Issue by BR33Hacker** [BR33Hacker/vitejs-vite-1cprfl#1]
import pandas as pd
from surprise import Dataset, Reader, SVD
from surprise.model_selection import train_test_split
from surprise.accuracy import rmse

# Sample dataset
data = {
    "user_id": [1, 1, 1, 2, 2, 3, 3, 3, 4, 4],
    "movie_id": [101, 102, 103, 101, 104, 102, 103, 105, 101, 105],
    "rating": [5, 4, 3, 5, 4, 2, 3, 4, 4, 5],
}

# Convert to DataFrame
df = pd.DataFrame(data)

# Define the data reader
reader = Reader(rating_scale=(1, 5))
data = Dataset.load_from_df(df[["user_id", "movie_id", "rating"]], reader)

# Split data into train and test sets
trainset, testset = train_test_split(data, test_size=0.25, random_state=42)

# Build and train the recommendation model
model = SVD()
model.fit(trainset)

# Test the model
predictions = model.test(testset)
rmse(predictions)

# Function to recommend movies for a specific user
def recommend_movies(model, user_id, movie_ids, n_recommendations=5):
    """Generate movie recommendations for a user."""
    # Predict ratings for all movies the user hasn't rated
    predictions = [
        (movie_id, model.predict(user_id, movie_id).est)
        for movie_id in movie_ids
    ]
    # Sort by predicted rating in descending order
    recommendations = sorted(predictions, key=lambda x: x[1], reverse=True)
    return recommendations[:n_recommendations]

# List of all movies
all_movie_ids = df["movie_id"].unique()

# Example: Get recommendations for user 1
user_id = 1
rated_movies = df[df["user_id"] == user_id]["movie_id"].unique()
unrated_movies = [m for m in all_movie_ids if m not in rated_movies]

recommendations = recommend_movies(model, user_id, unrated_movies)
print(f"Recommendations for User {user_id}:")
for movie_id, predicted_rating in recommendations:
    print(f"Movie ID: {movie_id}, Predicted Rating: {predicted_rating:.2f}")
